### PR TITLE
Support multiple VS Code windows for files opened via Zowe Explorer

### PR DESCRIPTION
**Pull request by JillieBeanSim**
## Proposed changes

Note to testers - run integration clean up on master branch before moving to this branch and run integration setup on this branch.  I changed AllocLike integration test to fix other broken tests on master so it will remove it's data set after test vs cleanup script doing it.

Note # 2 This PR will need to be built, packaged, and installed for `Open New Window` to be tested.  May need to clean local cache of old install.

What Changed:
Removed the call to deactivate during activation and have deactivate check for the new Setting `zowe.files.temporaryDownloadsFolderCleanup` whose default is true.  When true, deactivate will clean up temp folder as ususal.  When false, deactivate will not remove temp folder or any of it's files.

image

tempFolder-setting-defTrue-final

## Release Notes

Milestone: 1.17.0

Changelog: Zowe Explorer cleans up the folders with file contents downloaded from z/OS during deactivation to not leave source code behind. If you open multiple VS Code windows to edit the same files side-by-side and then close one of these windows it would mean that file content gets deleted even if the file is still being edited in the other window. To prevent this kind of clean-up enable this setting. If you want to ensure that no source is left behind on a development machine then keep this setting disabled and only use one VS Code window at a time.

## Types of changes

What types of changes does your code introduce to Zowe Explorer?
_Put an `x` in the boxes that apply_

- [x] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Updates to Documentation or Tests (if none of the other choices apply)

## Checklist

_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This checklist will be used as reference for both the contributor and the reviewer_

- [x] I have read the [CONTRIBUTOR GUIDANCE](https://github.com/zowe/vscode-extension-for-zowe/wiki/Best-Practices:-Contributor-Guidance) wiki
- [x] PR title follows [Conventional Commits Guidelines](https://www.conventionalcommits.org/en/v1.0.0-beta.2/)
- [x] PR Description is included
- [x] gif or screenshot is included if visual changes are made
- [x] `yarn workspace vscode-extension-for-zowe vscode:prepublish` has been executed
- [x] All checks have passed (DCO, Jenkins and Code Coverage)
- [ ] I have added unit test and it is passing
- [x] I have added integration test and it is passing
- [x] There is coverage for the code that I have added
- [x] I have tested it manually and there are no regressions found
- [ ] I have added necessary documentation (if appropriate)
- [ ] Any PR dependencies have been merged and published (if appropriate)

## Further comments

If this is a relatively large or complex change, kick off the discussion by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

File: packages/zowe-explorer/src/extension.ts [hunk 1]
```diff
@@ -66,16 +66,22 @@ export async function activate(context: vscode.ExtensionContext): Promise<ZoweEx
...
-    // this is to handle if the application crashed on a previous execution and
-    // VSC didn't get a chance to call our deactivate to cleanup.
-    await deactivate();
-
     try {
-        fs.mkdirSync(globals.ZOWETEMPFOLDER);
-        fs.mkdirSync(globals.ZOWE_TMP_FOLDER);
-        fs.mkdirSync(globals.USS_DIR);
-        fs.mkdirSync(globals.DS_DIR);
+        if (!fs.existsSync(globals.ZOWETEMPFOLDER)) {
```
Comment (nickImbirev) [inline packages/zowe-explorer/src/extension.ts:70]: Just out of curiosity: why we are using the synchronous API to work with the file system?

Just a suggestion: Maybe we can use the [async API](https://code.visualstudio.com/api/references/vscode-api#FileSystem), which is built in VSCode?

The benefits will be: better user experience, better resources consumption and of course, this VSCode API will work fine in Theia, in Eclipse Che or in some browser VSCode versions with ZE installed. It will also handle potential problems between the operating systems, because it is basically another layer of abstraction above the usual file system API.

**Comment by JillieBeanSim** [comment]
All integration tests are fixed here except the Delete failure which is being handled in a different PR

File: packages/zowe-explorer/src/extension.ts [hunk 2]
```diff
@@ -66,16 +66,13 @@ export async function activate(context: vscode.ExtensionContext): Promise<ZoweEx
     // Determine the runtime framework to support special behavior for Theia
     globals.defineGlobals(preferencesTempPath);
 
-    // Call cleanTempDir before continuing
-    // this is to handle if the application crashed on a previous execution and
-    // VSC didn't get a chance to call our deactivate to cleanup.
-    await deactivate();
```
Comment (phaumer) [inline packages/zowe-explorer/src/extension.ts:72]: Oh wow, that was a bad idea. Why not call cleanup directly? What if others would have added more code to the deactiviate function that would have broken activation?

**Review by phaumer — CHANGES_REQUESTED** [review]
We can merge this with the requested changes, but I am wondering if our users would expect this to work automatically using reference counters?

File: packages/zowe-explorer/src/extension.ts (hunk 2 above)
Comment (JillieBeanSim) [reply packages/zowe-explorer/src/extension.ts:72]: @phaumer should I add back a call to cleanup?  calling cleanup during activation is what was causing the deletion of a file right when new window was opened.  With this PR that is not done now but if the new setting is set to false (default) once one of the new windows closes `deleted` will show in other open windows if a ZE file is opened.  If set to true, that doesn't happen since no cleanup is done on temp folder.

File: packages/zowe-explorer/package.nls.json [hunk 3]
```diff
@@ -73,6 +73,7 @@
   "Zowe-Default-Commands-Persistent-Favorites": "Toggle if Commands persist locally",
   "Zowe-Environment": "Environment where the extension is running, default is VSCode",
   "Zowe-Automatic-Validation": "Allow automatic validation of profiles.",
+  "Zowe-Disable-TempFolder-Cleanup": "Disable the temporary folder cleanup during deactivation.",
```
[the remark this answers is not in the document]
Comment (lauren-li) [reply packages/zowe-explorer/package.nls.json]: @phaumer @JillieBeanSim I think if we want to go with the sub-groups, it's fine to start doing that in this PR for the setting you are adding. We can go ahead and use `zowe.files.disableTemporaryDownloadsFolderCleanup` for this PR.

For other existing settings that would need to be migrated, I would leave those for @rudyflores' WIP that will handle the settings migration.

File: packages/zowe-explorer/package.json [hunk 4]
```diff
@@ -1652,6 +1652,12 @@
           },
           "description": "%Zowe-Environment%",
           "scope": "window"
+        },
+        "zowe.disableTemporaryDownloadFolderCleanup": {
+          "type": "boolean",
+          "default": false,
+          "description": "%zowe.disableTemporaryDownloadFolderCleanup%",
```
Comment (lauren-li) [inline packages/zowe-explorer/package.json:1659]: @JillieBeanSim Would it be less confusing for users if we labeled this setting in the positive? Saying "Disable Temporary Downloads Folder Cleanup" forces the user to think in double negatives when setting this to `false`, i.e. "I am not disabling `x`, so therefore I am enabling `x`".

Perhaps we could use one of the following:
- "Enable Temporary Downloads Folder Cleanup" (true/false)
- "Temporary Downloads Folder Cleanup"
  - Settings description: "Enable temporary downloads folder cleanup" (true/false).
  
I remember we had a similar discussion about double negatives when we added the global enable/disable profile validation setting, and this is what we ended up going with:

Screen Shot 2021-06-29 at 9 34 06 AM

Comment (JillieBeanSim) [reply packages/zowe-explorer/package.json:1659]: @lauren-li so we don't want to include enable or disable in the label, just an open ended label, and add Allow to the description like with validation?

Comment (JillieBeanSim) [reply packages/zowe-explorer/package.json:1659]: `zowe.files.temporaryDownloadsFolderCleanup`: `Allow deletion of downloaded file content at editor deactivation (closing a Window) to support using separate VS Code windows editing the same files.`

Comment (JillieBeanSim) [reply packages/zowe-explorer/package.json:1659]: image

Comment (JillieBeanSim) [reply packages/zowe-explorer/package.json:1659]: Sorry @lauren-li , the above examples are all wrong.

What we will need to do if we follow the validation example is change to this with default true and allow the user to disable it by unchecking the box.
image

Comment (lauren-li) [reply packages/zowe-explorer/package.json:1659]: @JillieBeanSim I would rephrase the setting description to:
```
Allow deletion of downloaded file content during deactivation (closing a window). Disable this setting to support using separate VS Code windows editing the same files.
```
Do we want to add a note that disabling the setting means the temporary files will persist after the window is closed?

Actually, would it be clearer if we renamed the setting to `Persist temporary downloads`? (The default would have to be set to `false` again, though.)

File: packages/zowe-explorer/src/extension.ts (hunk 2 above)
Comment (phaumer) [reply packages/zowe-explorer/src/extension.ts:72]: No, sorry, I was just surprised that we had that code. Should have not commented.

**Comment by JillieBeanSim** [comment]
as discussed, we will follow the validation example and change to this with default true and allow the user to disable it by unchecking the box.
image

I will update the PR description to match these changes

File: packages/zowe-explorer/src/utils/TempFolder.ts [hunk 5]
```diff
@@ -104,12 +105,16 @@ export function cleanDir(directory) {
  * @export
  */
 export async function cleanTempDir() {
+    // Get temp folder location from settings
```
Comment (lauren-li) [inline packages/zowe-explorer/src/utils/TempFolder.ts:108]: @JillieBeanSim The comment in this line doesn’t seem to describe what is actually happening in the block below it. Should it maybe say `// Get temp folder cleanup preference from settings`?

**Review by lauren-li — DISMISSED** [review]
I think this PR looks good. All unit and expected integration tests are passing. The new setting description makes clear the benefit of enabling versus disabling the temp folder cleanup, and toggling the setting on/off works as expected in manual testing.

The only (minor) question I had was about a comment (see below), but I don't think it's a blocker to this PR. I don't mind re-approving if edits are made to the comment.

Thanks @JillieBeanSim for adding this feature!

**Review by lauren-li — DISMISSED** [review]
This looks good to me now. Thanks @JillieBeanSim for addressing my comment!

**Comment by phaumer** [comment]
Should this comment still be addressed? https://github.com/zowe/vscode-extension-for-zowe/pull/1347#discussion_r647182685

**Comment by JillieBeanSim** [comment]
> Should this comment still be addressed? [#1347 (comment)](https://github.com/zowe/vscode-extension-for-zowe/pull/1347#discussion_r647182685)

@phaumer @nickImbirev when I tested with the VS Code APIs there was some misbehaviors, it may be why we are using the synchronous API to work with the file system.  Could be something to look further into as technical debt.

**Comment by lauren-li** [comment]
@JillieBeanSim @crawr @phaumer This PR refers to the temp folder as `TemporaryDownloadsFolder`, whereas PR #1373 refers to it as `TempFolder`. We should pick one way of referring to the temp folder and be consistent with using it.

This will also affect how @rudyflores renames the `Zowe-Temp-Folder-Location` setting for his upcoming PR.

**Comment by rudyflores** [comment]
After posting in the issue thread for #1263 the proposal for standardized names, it was decided to change the current `zowe.files.temporaryDownloadsFolderCleanup` to the option proposed by @VitGottwald: `zowe.files.temporaryDownloadsFolder.cleanup` in order to keep the configuration names consistent.

Could you change the current name to the one that Vit proposed @JillieBeanSim ?

**Review by jellypuno — DISMISSED** [review]
This works really well! Thank  you @JillieBeanSim

**Comment by JillieBeanSim** [comment]
Please re-review @lauren-li & @jellypuno.  I addressed comment https://github.com/zowe/vscode-extension-for-zowe/pull/1347#issuecomment-876724996

**Comment by JillieBeanSim** [comment]
This is what the settings look like now with the changes
image

**Review by jellypuno — APPROVED** [review]

**Review by lauren-li — APPROVED** [review]
This PR looks good to me. All unit and expected integration tests are passing, and manual testing still works well with the renamed setting. Thanks @JillieBeanSim for addressing all the comments!
